gap/oper.gi

```diff
@@ -2183,6 +2183,141 @@ function(D, root)
+# First, notice that the cycle space is composed of orthogonal subspaces
+# corresponding to the cycle spaces of the connected components.
+# e.g. if G has G1, G2 and G3 connected
+# components with B1, B2 and B3 cycle basis matrix respectively, then the
+# resulting cycle basis matrix of G is
+# [[ B1,  0,  0],
+#  [  0, B2,  0],
+#  [  0,  0, B3]]
+# up to some permutation on the order of the edges.
+# As a result, we can compute the fundamental cycle basis of each connected
+# component and then combine them.
+
+# For each connected component, a spanning tree is computed rooted at 1. Then,
+# there is one to one correspondence between the edges not in the spanning tree
+# and the fundamental cycles basis.
+# (See : https://en.wikipedia.org/wiki/Cycle_basis#Fundamental_cycles)
+# The set of edges that form the base cycle is computed by finding the path
+# from the root to the each sides of the edge and then adding the edge to the
+# path. Then, it is converted to a binary vector where the i-th entry is 1 if
+# the i-th edge in the 'EdgesList' is in the cycle and 0 otherwise.
+# Related paper : https://dl.acm.org/doi/pdf/10.1145/363219.363232
+InstallMethod(DigraphCycleBasis, "for a digraph",
+[IsDigraph],
+function(G)
+  local OutNbr, InNbr, n, partialSum, m, roots, visited, unusedEdges, i, c, s,
+    z, queue, u, v, p, B;
+
+  # Check for loops
+  if DigraphHasLoops(G) then
+    ErrorNoReturn("the 1st argument (a digraph) must not have any loops");
+  fi;
+
+  G := MaximalAntiSymmetricSubdigraph(G);
+  OutNbr := OutNeighbors(G);
+  InNbr := InNeighbors(G);
+  Unbind(G);
+
+  n := Length(OutNbr);
+
+  # Quick early return for too few vertices
+  if n < 3 then
+    return [OutNbr, []];
+  fi;
+  # Find the partial sum of each row of OutNbr
+  partialSum := [0];
+  for i in [1 .. n - 1] do
+    Add(partialSum, partialSum[i] + Length(OutNbr[i]));
+  od;
+  m := partialSum[n] + Length(OutNbr[n]);
+  # Quick early return for too few edges
+  if m < 3 then
+    return [OutNbr, []];
+  fi;
+
+  # Traverse the graph, depth first search
+  roots := [];
+  visited := BlistList([1 .. n], []);
+  unusedEdges := [];
+  while ForAny(visited, x -> x = false) do
+    s := Position(visited, false);
+    Add(roots, s);
+    visited[s] := [0, s];
+    queue := [s];
+    while not IsEmpty(queue) do
+      u := Remove(queue);
+      for p in [1 .. Length(OutNbr[u])] do
+        v := OutNbr[u][p];
+        i := partialSum[u] + p;
+        if visited[v] = false then
+          visited[v] := [i, u];
+          Add(queue, v);
+        elif v in queue then
+          Add(unusedEdges, [u, i, v, visited[v][1], visited[v][2]]);
+        fi;
+      od;
+      for v in InNbr[u] do
+        p := Position(OutNbr[v], u);
+        i := partialSum[v] + p;
+        if visited[v] = false then
+          visited[v] := [i, u];
+          Add(queue, v);
+        elif v in queue then
+          Add(unusedEdges, [u, i, v, visited[v][1], visited[v][2]]);
```

Comment:
Seems like `v` isn't used later, so please remove it, and adjust the later code too.